### PR TITLE
[Lang] Support sparse matrix datatype and storage format configuration

### PR DESCRIPTION
**Pull request by FantasyVR**
Related issue = #2906

This PR seems messy for the reviewer. 
For short, the eigen sparse matrix is warpped using `EigenSparseMatrix` class which is derived from the abstract class `SparseMatrix`. 
The data type and storage format is configured using `SparseMatrixBuilder `interface. The `EigenSparseMatrix` is created and built from triplets in `SparseMatrixBuilder`.

### CODE REVIEW

File: tests/python/test_sparse_matrix.py [hunk 1]
```diff
@@ -4,17 +4,21 @@
...
     Abuilder = ti.linalg.SparseMatrixBuilder(n,
                                              n,
                                              max_num_triplets=100,
-                                             dtype=dtype)
+                                             dtype=dtype,
+                                             storage_format=storage_format)
 
     @ti.kernel
-    def fill(Abuilder: ti.types.sparse_matrix_builder()):
+    def fill(Abuilder: ti.linalg.sparse_matrix_builder()):
```
Comment (strongoier) [inline tests/python/test_sparse_matrix.py:21]:
```suggestion
    def fill(Abuilder: ti.types.sparse_matrix_builder()):
```
This shouldn't be changed.

File: taichi/program/sparse_matrix.h [hunk 2]
```diff
@@ -36,37 +40,153 @@ class SparseMatrixBuilder {
...
+  EigenSparseMatrix(EigenSparseMatrix &&sm)
+      : SparseMatrix(sm.num_rows(), sm.num_cols(), sm.dtype_),
+        matrix_(sm.matrix_) {
+  }
+  explicit EigenSparseMatrix(const EigenMatrix &em)
+      : SparseMatrix(em.rows(), em.cols()), matrix_(em) {
+  }
+
+  virtual ~EigenSparseMatrix() override = default;
+  virtual void build_triplets(void *triplets_adr) override;
```
Comment (k-ye) [inline taichi/program/sparse_matrix.h:114]: No need to add `virtual` for the **inherited methods**, just `override`

File: taichi/program/sparse_matrix.h [hunk 3]
```diff
@@ -36,37 +40,153 @@ class SparseMatrixBuilder {
...
+  }
+  explicit EigenSparseMatrix(EigenSparseMatrix &&sm)
+      : SparseMatrix(sm.num_rows(), sm.num_cols(), sm.dtype_),
+        matrix_(sm.matrix_) {
+  }
+  explicit EigenSparseMatrix(const EigenMatrix &em)
+      : SparseMatrix(em.rows(), em.cols()), matrix_(em) {
+  }
+
+  virtual ~EigenSparseMatrix() = default;
```
Comment (strongoier) [inline taichi/program/sparse_matrix.h:113]:
```suggestion
~EigenSparseMatrix() override = default;
```

**Review by strongoier — APPROVED** [review]
LGTM!

**Pull request merged**